notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
                 'app': 'notebook-worker',
                 'hail.is/notebook-instance': INSTANCE_ID,
                 'uuid': pod_id,
-            }),
+                'svc_name': svc.metadata.name,
+                **labels
+            },),
         spec=pod_spec)
     pod = k8s.create_namespaced_pod(
         'default',
         pod_template,
-        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS,
+    )
+
     return svc, pod
 
 
-def external_url_for(path):
-    # NOTE: nginx strips https and sets X-Forwarded-Proto: https, but
-    # it is not used by request.url or url_for, so rewrite the url and
-    # set _scheme='https' explicitly.
-    protocol = request.headers.get('X-Forwarded-Proto', None)
-    url = flask.url_for('root', _scheme=protocol, _external='true')
-    return url + path
+def del_pod(pod_name: str):
+    k8s.delete_namespaced_pod(
+        pod_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+def del_svc(svc_name: str):
+    k8s.delete_namespaced_service(
+        svc_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+# General resource marshalling functions #
+
+
+def get_path(data, path: str):
+    # Not using tail recursion because python doesn't optimize such calls
+    while True:
+        idx = path.find('.')
+
+        if idx == -1:
+            if isinstance(data, dict):
+                return data[path]
+            return getattr(data, path)
+
+        data = getattr(data, path[0:idx])
 
+        path = path[idx + 1:]
+
+
+def marshall_json(resources: [], paths=[], flatten=False):
+    if len(resources) == 0 or len(paths) == 0:
+        if flatten:
+            return "{}"
+
+        return "[]"
+
+    resp = []
+    for rsc in resources:
+        data = {}
+        for path in paths:
+            if len(path) == 3:
+                data[path[0]] = path[2](get_path(rsc, path[1]))
+            else:
+                data[path[0]] = get_path(rsc, path[1])
+
+        resp.append(data)
+
+    if flatten and len(resources) == 1:
+        return ujson.dumps(resp[0])
+
+    return ujson.dumps(resp)
+
+
+def read_svc_status(svc_name: str):
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
```

Comment:
This name is a bit confusing since it converts the first container_status's state to a dictionary and returns that. Also, V1ContainerStatus and V1ContainerState have a `to_dict()`. I'm inclined to replace uses of this with `statuses[0].state.to_dict`. We always expect exactly one container (we ensure that in this file when we construct pods), so I don't see any reason to handle other cases.

Reply:
I want to dictate the exact shape of the response to the user.